### DISCUSSION TITLE
Is it possible to add simulators to a group outside of a group?

### DISCUSSION THREAD

**Discussion by bl0ckhaT** [OFFIS-mosaik/mosaik#19] [Q&A] [answered]
Hello,

I am using multiple simulators, which are configured and started at different parts of the co-simulation. When reading the documentation, I didn't find any other way of adding simulators to a group. Hence my question.

If it isn't possible, I would like to know why. I am not very familiar with the "with" statement, I only know that it works with context managers.

To anyone taking the time to read this, thank you 🙏 .

**Comment by eike** [comment, marked as answer]
This is currently not possible, no. (Well, it's Python, so you could probably find a hacky way to do it: The `with` statement does not do anything except replace the `World`'s `current_group` by a subgroup, execute the statements within the `with` statement, and then replace the subgroup by its parent again. If you carefully replace `current_group` yourself, things "should" still work. Be advised that we don't consider `current_group` to be a part of the public API, so future mosaik version might change how it works. In fact, in the development version, `World` delegates its work to an `AsyncWorld`, and you would have to modify `current_group` on the `AsyncWorld`, instead.)

The point of using the `with` statement was to ensure that the groups are properly nested already at the syntactic level. Our impression is that usually, simulators that would share a group would also be started somewhat close together in the scenario. Even if your simulators are started by different methods, this is not automatically impossible. As an example:

```python
class SimAStarter:
    def __init__(self, world):
        self.world = world

    def start(self):
        self.world.start("SimA")

class SimBStarter:
    # similar to SimAStarter
    ...

def main():
    world = World(SIM_CONFIG)
    sim_a_starter = SimAStarter(world)
    sim_b_starter = SimBStarter(world)

    with world.group():
        sim_a_starter.start()
        sim_b_starter.start()
```

Even though the actual `world.start` calls are not physically within the `with` block, they're called from it, and therefore executed within its context.

That said, I am willing to reconsider the strictness of our approach for a future version of mosaik if it turns out that reasonable scenarios are difficult to implement with the current one.

**Reply by bl0ckhaT** [reply]
Thank you. Will try to solve it with your idea.

On another note: 
Do you have an example of how a correct instantiation of a SimGroup looks like?

- Also, does the connecting of attributes and entities need to happen within the group() contextmanager? If so, why? I'm asking, since the group disappears after the "with" expression is exited, which is the expected behavior as you outlined above.
- Lastly, how does Mosaik then know which simulators are in the same group, if the group only exists within the contextmanager?

**Reply by eike** [reply]
The `SimGroup`s in a simulation should form a tree with each `SimGroup` pointing to a `parent` except for the `main_group` of `AsyncWorld`, which has `None` as a parent. When a simulator is started, it takes the world's `current_group` for the `_group` field of its `AsyncModelFactory` and the depth (i.e. the number of steps to get to the main group) is also given to the simulator's `SimRunner`. The world's `current_group` starts out as the `main_group`, and a `with world.group()` block establishes a new subgroup of the current group (i.e. one with `parent` set to the current group) and then uses this subgroup as the current group for the duration of that block.

If you really want, you can create `SimGroup`s manually (by just calling the constructor with the parent group as an argument). To get simulators to use this group, you can then either update the `AsyncWorld`'s `current_group` before calling `world.start`. Changing the `_group` of an `AsyncModelFactory` afterwards will almost certainly not work, as the `SimRunner` will have already done things based on the previous depth.

To you additional questions:

- No, connections don't need to be made in the ``with`` block (but they can, if you find this more readable).
- The group is attached to simulators started in the ``with`` block (as `AsyncModelFactory._group`). When connecting entities, the groups search their joint parent to determine how to handle weak connections.
